# Part 1: Data Understanding and Preparation

1. First, I loaded the dataset. Then, in order for the file to read the dataset, I moved the file into /content working directory.

2. Each row represents an individual insurance customer. The row includes the customer's age, sex, BMI, number of children, whether the customer is a smoker or not, the region they're located in, and the cost of charges.

3. The target variable is "charges." This dependent variable represents the medical charges for the customer, which is an outcome of the other input values in the dataset.

4. After inspecting the dataset's data types, missing values, and anomalies, there are no issues with any of the aforementioned.

In [53]:
import pandas as pd # importing pandas library and creating an alias of pandas called pd

df = pd.read_csv("insurance.csv") # reading "insurance.csv" using pandas library

y = df["charges"]          # extracts the target variable, "charges"
X = df.drop("charges", axis=1)   # removes "charges" from the dataset for the WHOLE COLUMN (axis=1)

In [54]:
# One-hot encoding
# Converts variable into binary indicator variables
# get_dummies is used for "converting categorical variables into dummy/indicator variables"
# if value is represented then true, else false
# Using drop_first=True to drop 1 table and use as a reference baseline. Prevents perfect multicollinearity
X_encoded = pd.get_dummies(X, columns=["sex", "smoker", "region"], drop_first=True)

In [55]:
X_encoded = X_encoded.astype(int) # casts the data as an int

In [56]:
print(X_encoded.shape) # .shape for row and column count
print(y.shape)

# Output
# (1338, 8) --> 1338 samples, 8 predictor variables
# (1338,) --> 1 target

(1338, 8)
(1338,)


# Part 2: Supervised Model (Regression)
Fit an Ordinary Least Squares (OLS) regression model using the prepared data.
Purpose: to predict the target variable, "charges."

In [57]:
import statsmodels.api as sm # For statistical modeling

X_ols = sm.add_constant(X_encoded)
model = sm.OLS(y, X_ols).fit()

In [58]:
# Sanity Check
# Ensures that any changes haven't broken the code
X_ols.dtypes #dtypes returns the data types

,0
const,float64
age,int64
bmi,int64
children,int64
sex_male,int64
smoker_yes,int64
region_northwest,int64
region_southeast,int64
region_southwest,int64


In [59]:
# Overall summary of the OLS Regression Results
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                charges   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.749
Method:                 Least Squares   F-statistic:                     500.7
Date:                Mon, 16 Feb 2026   Prob (F-statistic):               0.00
Time:                        20:18:21   Log-Likelihood:                -13548.
No. Observations:                1338   AIC:                         2.711e+04
Df Residuals:                    1329   BIC:                         2.716e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const            -1.174e+04    975.287    -12.039      0.000   -1.37e+04   -9828.163
age                256.9769     11.898     21.597      0.000     233.635     280.319
bmi                337.9889     28.516     11.853      0.000     282.048     393.930
children           478.2260    137.810      3.470      0.001     207.878     748.574
sex_male          -132.4530    332.970     -0.398      0.691    -785.658     520.752
smoker_yes        2.386e+04    413.193     57.745      0.000     2.3e+04    2.47e+04
region_northwest  -346.4336    476.304     -0.727      0.467   -1280.823     587.956
region_southeast -1042.4067    478.895     -2.177      0.030   -1981.879    -102.935
region_southwest  -969.5561    478.032     -2.028      0.043   -1907.335     -31.777
==============================================================================
Omnibus:                      299.849   Durbin-Watson:                   2.089
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              716.614
Skew:                           1.210   Prob(JB):                    2.45e-156
Kurtosis:                       5.646   Cond. No.                         306.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [60]:
# Report estimated numeric coefficients for each variable
model.params

,0
const,-11741.432494
age,256.976935
bmi,337.988864
children,478.225974
sex_male,-132.453026
smoker_yes,23859.788592
region_northwest,-346.433637
region_southeast,-1042.406732
region_southwest,-969.556124


In [61]:
# Report r-squared
# Purpose of r-squared: percentage of variance of charges explained by the model
print("r-squared: ", model.rsquared)

r-squared:  0.7508831656660024


## Interpret at least two coefficients in plain language

- Smoker coefficient:
The smoker coefficient (smoker_yes) is approximately 23859.78. This means that as long as the other variables remain the same, being a smoker raises the expected medical charges by approximately $23,859.79.

- BMI coefficient:
The BMI coefficient is approximately 337.99. This means that if the BMI increases by one point, the expected medical charges will increase by approximately $337.99.

## Explain what the model claims about the relationship between inputs and the target

Based off the OLS Regression Results, the model is claiming that smoking contributes the most towards a higher medical cost. The model also claims that variables like age and BMI have positive effects towards the overall medical cost. So, being older or having a higher BMI will increase a customer's medical cost. Variables like the number of children and region have little to no influence on the overall medical cost.

# Sources/Citations
Importing pandas library:
https://www.w3schools.com/python/pandas/pandas_getting_started.asp

Reading files using pandas:
https://www.w3schools.com/python/pandas/pandas_csv.asp

One-hot encoding:
https://www.geeksforgeeks.org/machine-learning/ml-one-hot-encoding/

get_dummies:
https://www.geeksforgeeks.org/pandas/python-pandas-get_dummies-method/

Dropping columns:
https://stackoverflow.com/questions/63661560/drop-first-true-during-dummy-variable-creation-in-pandas

Data types:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html

Sanity check:
https://www.reddit.com/r/QualityAssurance/comments/1dqpx39/can_someone_please_explain_the_difference_between/

.shape (Sanity check):
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html

Casting (.astype):
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html

OLS Model in Python:
https://www.youtube.com/watch?v=bw_wBbUKJW8

Importing statsmodels.api module:
https://www.statsmodels.org/stable/index.html

Reporting the estimated coefficients:
https://www.geeksforgeeks.org/data-analysis/extracting-regression-coefficients-from-statsmodelsapi/

R-squared:
https://www.reddit.com/r/statistics/comments/1cwvp5w/q_how_to_interpret_an_r2_value/#:~:text=R2%20is%20the%20percentage,predicted%20values%20were%20perfectly%20accurate.